### PR TITLE
[SYCL][ESIMD][EMU] Atomic update

### PR DESCRIPTION
**Pull request by dongkyunahn-intel**
- __esimd_svm_atomic0/1/2
- cmpxchng argument order fix (New value first, expected value second)
- atomic_add/sub/min/max/cmpxchg update

### CODE REVIEW

**Comment by dongkyunahn-intel** [comment]
'dword_atomic_smoke.cpp' from intel/llvm-test-suite fails with infinite looping. Still debugging.

File: sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp [hunk 1]
```diff
@@ -39,7 +45,21 @@ template <typename Ty> Ty atomic_add_fetch(Ty *ptr, Ty val) {
...
+    return __atomic_add_fetch(ptr, val, __ATOMIC_SEQ_CST);
+  } else {
+    // For Floating type
+    Ty _old, _new;
+    CmpxchgTy<Ty> _old_bits, _new_bits;
+    do {
+      _old = *ptr;
+      _new = _old + val;
+      _old_bits = *(CmpxchgTy<Ty> *)&_old;
+      _new_bits = *(CmpxchgTy<Ty> *)&_new;
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp:58]: Nit: these two can be moved out of the loop (maybe in a separate patch)

**Comment by dongkyunahn-intel** [comment]
> 'dword_atomic_smoke.cpp' from intel/llvm-test-suite fails with infinite looping. Still debugging.

The test is disabled in intel/llvm-test-suite as the test fails for GPU as well - https://github.com/intel/llvm-test-suite/pull/1185

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp [hunk 2]
```diff
@@ -528,7 +588,27 @@ __esimd_svm_atomic2(__ESIMD_DNS::vector_type_t<uint64_t, N> addrs,
...
+    retv[AddrIdx] = *((Ty *)addrs[AddrIdx]);
+    if constexpr (Op == __ESIMD_NS::atomic_op::cmpxchg) {
+      __ESIMD_DNS::atomic_cmpxchg((Ty *)addrs[AddrIdx], src0[AddrIdx],
+                                  src1[AddrIdx]);
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::fcmpwr) {
+      if constexpr (__ESIMD_DNS::is_fp_type<Ty>::value) {
+        __ESIMD_DNS::atomic_cmpxchg((Ty *)addrs[AddrIdx], src0[AddrIdx],
+                                    src1[AddrIdx]);
+      }
+    }
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:609]:
```suggestion
    static_assert((Op == __ESIMD_NS::atomic_op::cmpxchg) || (Op == __ESIMD_NS::atomic_op::fcmpxchg));
    __ESIMD_DNS::atomic_cmpxchg((Ty *)addrs[AddrIdx], src0[AddrIdx], src1[AddrIdx]);
```

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp [hunk 3]
```diff
@@ -500,17 +518,59 @@ __esimd_svm_atomic1(__ESIMD_DNS::vector_type_t<uint64_t, N> addrs,
...
+        __ESIMD_DNS::atomic_max<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+      }
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::bit_and) {
+      // TODO : Type Check? Integral type only?
+      __ESIMD_DNS::atomic_and_fetch<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::bit_or) {
+      // TODO : Type Check? Integral type only?
+      __ESIMD_DNS::atomic_or_fetch<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::bit_xor) {
+      // TODO : Type Check? Integral type only?
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:572]: yes, integral only

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp [hunk 4]
```diff
@@ -500,17 +518,59 @@ __esimd_svm_atomic1(__ESIMD_DNS::vector_type_t<uint64_t, N> addrs,
...
+      __ESIMD_DNS::atomic_max<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::min) {
+      if constexpr (!__ESIMD_DNS::is_fp_type<Ty>::value) {
+        __ESIMD_DNS::atomic_min<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+      }
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::max) {
+      if constexpr (!__ESIMD_DNS::is_fp_type<Ty>::value) {
+        __ESIMD_DNS::atomic_max<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+      }
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::fadd) {
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:549]: fadd and integer add, fsub and integer sub, fmax and umax/smax, fmin and umin/smin cases can be merged

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp [hunk 5]
```diff
@@ -485,7 +485,25 @@ __esimd_svm_atomic0(__ESIMD_DNS::vector_type_t<uint64_t, N> addrs,
...
+    }
+    // Keeping original values for return + 'load'
+    Output[AddrIdx] = *((Ty *)addrs[AddrIdx]);
+    if constexpr (Op == __ESIMD_NS::atomic_op::inc) {
+      __ESIMD_DNS::atomic_add_fetch<Ty>((Ty *)addrs[AddrIdx],
+                                        static_cast<Ty>(1));
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::dec) {
+      __ESIMD_DNS::atomic_sub_fetch<Ty>((Ty *)addrs[AddrIdx],
+                                        static_cast<Ty>(1));
+    }
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:504]: This is wrong. The update becomes non-atomic. There can be other updates happening between writing to Output above and the actual atomic operation.
```suggestion
    if constexpr (Op == __ESIMD_NS::atomic_op::inc) {
      Output[AddrIdx] = __ESIMD_DNS::atomic_add_fetch<Ty>((Ty *)addrs[AddrIdx],
                                        static_cast<Ty>(1));
    } else if constexpr (Op == __ESIMD_NS::atomic_op::dec) {
      Output[AddrIdx] = __ESIMD_DNS::atomic_sub_fetch<Ty>((Ty *)addrs[AddrIdx],
                                        static_cast<Ty>(1));
    }
```

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp [hunk 6]
```diff
@@ -500,17 +518,59 @@ __esimd_svm_atomic1(__ESIMD_DNS::vector_type_t<uint64_t, N> addrs,
...
 
-      switch (Op) {
-      case __ESIMD_NS::atomic_op::add:
-        retv[i] = __ESIMD_DNS::atomic_add_fetch<Ty>(p, src0[i]);
-        break;
-      default:
-        __ESIMD_UNSUPPORTED_ON_HOST;
+    // Keeping original values for return
+    retv[AddrIdx] = *((Ty *)addrs[AddrIdx]);
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:529]: this is wrong (see above [comment](https://github.com/intel/llvm/pull/6661/files#r958778500))

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp [hunk 7]
```diff
@@ -528,7 +588,27 @@ __esimd_svm_atomic2(__ESIMD_DNS::vector_type_t<uint64_t, N> addrs,
...
+  __ESIMD_DNS::vector_type_t<Ty, N> retv;
+
+  for (int AddrIdx = 0; AddrIdx < N; AddrIdx++) {
+    if (pred[AddrIdx] == 0) {
+      // Skip Output vector elements correpsonding to
+      // predicates whose value is zero
+      continue;
+    }
+    // Keeping original values for return
+    retv[AddrIdx] = *((Ty *)addrs[AddrIdx]);
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:600]: this is wrong, see above [comment](https://github.com/intel/llvm/pull/6661/files#r958778500)

File: sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp (hunk 1 above)
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp:60]: Will do later in another PR.

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp (hunk 3 above)
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp]: Integral type checks are applied in `__ESIMD_DNS::atomic_*` functions.

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp (hunk 4 above)
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp]: Done.

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp (hunk 6 above)
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp]: Done. Return values from `__atomic_*` functions are used.

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp (hunk 7 above)
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp]: Done. Return values from `__atomic_*` functions are used.

File: sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp [hunk 8]
```diff
@@ -16,12 +16,18 @@ namespace __ESIMD_DNS {
...
+// 'bridging' variables in integral type.
+template <typename Ty> using CmpxchgTy = __ESIMD_DNS::uint_type_t<sizeof(Ty)>;
+
 template <typename Ty> Ty atomic_load(Ty *ptr) {
 #ifdef _WIN32
   // TODO: Windows will be supported soon
   __ESIMD_UNSUPPORTED_ON_HOST;
 #else
-  return __atomic_load(ptr, __ATOMIC_SEQ_CST);
+  return __atomic_load_n((CmpxchgTy<Ty> *)ptr, __ATOMIC_SEQ_CST);
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp:30]: `type __atomic_load_n (type *ptr, int memorder)`
so AFAIU `return __atomic_load_n((CmpxchgTy<Ty> *)ptr, __ATOMIC_SEQ_CST);` does type conversion from CmpxchgTy<Ty> and Ty, which is wrong. Do you have tests for this? They should have caught this problem.
```suggestion
  return sycl::bit_cast<Ty>(__atomic_load_n((CmpxchgTy<Ty> *)ptr, __ATOMIC_SEQ_CST));
```

Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp:30]: I don't have a unit test for this. I'll put TODO comments with `__ESIMD_UNSUPPORTED_ON_HOST`.

**Comment by dongkyunahn-intel** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1259

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp [hunk 9]
```diff
@@ -1879,7 +1808,9 @@ __ESIMD_INTRIN void __esimd_lsc_fence(__ESIMD_DNS::simd_mask_storage_t<N> pred)
     ;
 #else  // __SYCL_DEVICE_ONLY__
 {
-  __ESIMD_UNSUPPORTED_ON_HOST;
+  // In ESIMD_EMULATOR device interface, write operations are applied
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:1811]: This is extremely dangerous, and can be a source of obscure bugs.
Emulator uses multiple threads to emulate GPU threads, so memory fence can't be a no-op.

**Comment by dongkyunahn-intel** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1274

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp (hunk 9 above)
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp]: Fixed - built-in `__atomic_thread_fence()` is called.

**Comment by dongkyunahn-intel** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1274

**Comment by dongkyunahn-intel** [comment]
Failures from ESIMD_EMULATOR are already handled in https://github.com/intel/llvm-test-suite/pull/1274

```
********************
Timed Out Tests (1):
  SYCL :: dword_atomic_cmpxchg.cpp

********************
Failed Tests (3):
  SYCL :: lsc/lsc_slm.cpp
  SYCL :: lsc/lsc_surf.cpp
  SYCL :: lsc/lsc_usm.cpp

********************
Unexpectedly Passed Tests (1):
  SYCL :: dword_atomic_smoke.cpp
```

File: sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp [hunk 10]
```diff
@@ -30,52 +39,87 @@ template <typename Ty> Ty atomic_store(Ty *ptr, Ty val) {
   // TODO: Windows will be supported soon
   __ESIMD_UNSUPPORTED_ON_HOST;
 #else
-  __atomic_store(ptr, val, __ATOMIC_SEQ_CST);
+  Ty ret = atomic_load<Ty>((CmpxchgTy<Ty> *)ptr);
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp:42]: it looks strange that we need atomic_load as a part of atomic_store implementation, but that's how Gen ISA is :)
no action for this comment is needed

File: sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp [hunk 11]
```diff
@@ -30,52 +39,87 @@ template <typename Ty> Ty atomic_store(Ty *ptr, Ty val) {
...
+  __atomic_store_n((CmpxchgTy<Ty> *)ptr, val, __ATOMIC_SEQ_CST);
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp:43]: Nit: strictly speaking, reinterpret_cast should be used. Can be fixed later.

File: sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp [hunk 12]
```diff
@@ -104,38 +157,64 @@ template <typename Ty> Ty atomic_max(Ty *ptr, Ty val) {
...
-    return _new;
+    return _old;
   } else {
-    __ESIMD_UNSUPPORTED_ON_HOST;
+    Ty _old, _new;
+    CmpxchgTy<Ty> _old_bits, _new_bits;
+    do {
+      _old = *ptr;
+      _new = std::max(_old, val);
+      _old_bits = *(CmpxchgTy<Ty> *)&_old;
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp:174]: Nit: sycl::bitcast should be used instead. Can be fixed later

File: sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp [hunk 13]
```diff
@@ -104,38 +157,64 @@ template <typename Ty> Ty atomic_max(Ty *ptr, Ty val) {
...
+    CmpxchgTy<Ty> desired_bits = *(CmpxchgTy<Ty> *)&desired;
+    CmpxchgTy<Ty> local_bits = *(CmpxchgTy<Ty> *)&expected;
+    __atomic_compare_exchange_n((CmpxchgTy<Ty> *)ptr, &local_bits, desired_bits,
+                                false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+    return *((Ty *)&local_bits);
   }
 #endif
 }
 
+// Template argument is to avoid failure from 'odr.cpp' test
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp:208]: I think this is a bad interface - using template parameter, please fix.

If ODR fails, a fix should be different.

Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp]: Template argument is replaced with `inline`.

**Comment by dongkyunahn-intel** [comment]
lsc_slm/surf/usm will be fixed once argument order for cmpxchg is confirmed.

Other failures are handled in https://github.com/intel/llvm-test-suite/pull/1274.
```
********************
Timed Out Tests (1):
  SYCL :: dword_atomic_cmpxchg.cpp

********************
Failed Tests (3):
  SYCL :: lsc/lsc_slm.cpp
  SYCL :: lsc/lsc_surf.cpp
  SYCL :: lsc/lsc_usm.cpp

********************
Unexpectedly Passed Tests (1):
  SYCL :: dword_atomic_smoke.cpp
```

File: sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp (hunk 10 above)
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp:42]: I asked Victor Mustya for clarification.

**Comment by dongkyunahn-intel** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1274

**Comment by dongkyunahn-intel** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1274

**Comment by dongkyunahn-intel** [comment]
Failures from 'SYCL / Linux / ESIMD Emu LLVM Test Suite (pull_request_target)'

```
********************
Timed Out Tests (1):
  SYCL :: dword_atomic_cmpxchg.cpp

********************
Unexpectedly Passed Tests (1):
  SYCL :: dword_atomic_smoke.cpp
```

These tests are modified in https://github.com/intel/llvm-test-suite/pull/1274

**Review by kbobrovs — APPROVED** [review]

**Pull request merged**